### DISCUSSION TITLE
Proposed coding Standards update

### DISCUSSION THREAD

**Discussion by sithlord48** [deskflow/deskflow#7762] [Development]
I would like to change our current naming rules  for Qt to this 

## Qt naming
1. Qt Controls SHALL NOT use  a `p` to indicate they are a pointer.
1. Qt Controls SHALL be named `<type><Description>` (`lblShortName`) 
                                                    OR `<type>_<lenghtyDescription>`  (`lbl_longerNamesMayUseThisStyle`)
1. Qt signals should indicate that something happened, e.g. `nameChanged`

### UI Form Specific Rules
1. Members of UI forms SHALL NOT use the `m_` prefix (i.e lblError NOT m_lblError) 
1. Don't use the autoconnection conventions (`on_foo_bar`) or `onFooBar`.
1. Ui Forms should always be in the namespace `Ui`
    1. Always use `ui{std::make_unique<Ui::Type>()}` make the ui instance
    1. The Ui Form SHALL be a private member of the class using
    1. The variable name SHALL be `ui` 
1. Ui Forms should never have connections internally 
1. When editing ui files take extra care to not accidentally change item sizes
1.When editing ui files take extra care to not accidentally change the current item on container based widgets.

### Qt Type Prefix scheme
  - `label`  QLabels that are  a static label
  - `lbl`      QLabels you will be changing programmatically
  - `rb`      QRadioButton
  - `cb`      QCheckBoxe
  - `combo`  QComboBox
  -  `btn`    Buttons (Tool or Push)
  -  `action`  QAction
  - Others name without the Q 

Justification: 

 - `Ui` is the standard namespace that ui files are in
 - `ui` is the standard name for ui files  (when used).
 -  All ui members have to used via the ui pointer
    - `m_pui->m_pLabelSomething` is harder to read then `ui->lblSomething` and uses more characters.
- autoconnections are error prone
- connections created in ui files are 
   - Harder to discover
   - Easier to duplicate
- All Qt Controls and most QObject will have to be pointer so the `p` is redundant in names.

**Comment by nbolton** [comment]
Cool. I'm mobile so I'll take a proper look soon but at first glance looks reasonable. Crack on with that I'd say. We should work toward implementing this piecemeal instead of big bang. I may do a few refactor PRs to bring the code up to the code style.

I also think we should try to get the GUI style and the Core style as close as possible so it's less confusing switching between the two.

**Comment by nbolton** [comment]
Thinking about what to put in the [code style doc](https://github.com/deskflow/deskflow/wiki/Code-Style), looks like you're suggesting we should keep `m_` for everything except `ui`.

If so, I think that makes sense because `ui` is always a member, making the `m_` prefix redundant for that member. I think we should describe this as a _special exception_ so people (who aren't Qt experts) don't get confused when they see the lack of `m_` (and think we don't use `m_`).

It seems like we also need a special exception for `Ui` namespace.

**Reply by sithlord48** [reply]
However you want to note it that works for me 
also a `p` exception for things Qt objects in general for ui items for sure as they are all member pointers.

**Comment by sithlord48** [comment]
ping on this can we put this into the code standards for Qt ?

**Reply by nbolton** [reply]
I think so. Are you looking for a volunteer or do you plan to do it? :)

**Reply by sithlord48** [reply]
yeah i can move this to the wiki
